crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -0,0 +1,1356 @@
+                    $( $accumulator )*
+                    $chunk_index += 1;
+                }
+            )
+        };
+
+        ( @_ [ $iterator:ident, $chunk_index:ident, $item_index:ident ] { [ $( $item:expr ),+ ] $( $rest:tt )* } { $( $accumulator:tt )* } ) => {
+            assert_items_eq!(
+                @_
+                [ $iterator, $chunk_index, $item_index ]
+                { $( $rest )* }
+                {
+                    $( $accumulator )*
+                    let expected_chunk_identifier = $iterator .peek().unwrap().1.chunk_identifier();
+                    $(
+                        assert_matches!(
+                            $iterator .next(),
+                            Some((chunk_index, ItemPosition(chunk_identifier, item_index), & $item )) => {
+                                // Ensure the chunk index (from the enumeration) is correct.
+                                assert_eq!(chunk_index, $chunk_index);
+                                // Ensure the chunk identifier is the same for all items in this chunk.
+                                assert_eq!(chunk_identifier, expected_chunk_identifier);
+                                // Ensure the item has the expected position.
+                                assert_eq!(item_index, $item_index);
+                            }
+                        );
+                        $item_index += 1;
+                    )+
+                    $item_index = 0;
+                    $chunk_index += 1;
+                }
+            )
+        };
+
+        ( @_ [ $iterator:ident, $chunk_index:ident, $item_index:ident ] {} { $( $accumulator:tt )* } ) => {
+            {
+                let mut $chunk_index = 0;
+                let mut $item_index = 0;
+                $( $accumulator )*
+            }
+        };
+
+        ( $linked_chunk:expr, $( $all:tt )* ) => {
+            assert_items_eq!(
+                @_
+                [ iterator, _chunk_index, _item_index ]
+                { $( $all )* }
+                {
+                    let mut iterator = $linked_chunk
+                        .rchunks()
+                        .enumerate()
+                        .filter_map(|(chunk_index, chunk)| match &chunk.content {
+                            ChunkContent::Gap => None,
+                            ChunkContent::Items(items) => {
+                                Some(items.iter().rev().enumerate().map(move |(item_index, item)| {
+                                    (chunk_index, ItemPosition(chunk.identifier(), item_index), item)
+                                }))
+                            }
+                        })
+                        .flatten()
+                        .peekable();
+                }
+            )
+        }
+    }
+
+    #[test]
+    fn test_chunk_identifier_generator() {
+        let generator = ChunkIdentifierGenerator::new_from_scratch();
+
+        assert_eq!(generator.generate_next(), Ok(1));
```

Comment:
YES.
No, but when initializing the `ChunkIdentifierGenerator`, we need a start value. Since it's a `u64`, I've used `0`. So the next is always 1. Not a big deal. It's identifier, not index 😉.